### PR TITLE
Created file in background

### PR DESCRIPTION
**Pull request by shiraji**
see #3 

I was just guessing my assumption at #3 is correct. If not, I will close this pull request.

~~Could you give me how to test this library? I haven't build/test this code, yet.~~ Didn't look at circle.yml. I will test this code in my local machine.

### CODE REVIEW

**Comment by shiraji** [comment]
Nice, it passes circle ci script. 

Now, questions are following:
- [Duration](https://github.com/takahirom/PreLollipopTransition/pull/18/files#diff-5608e74ca5914ec939cde04c740ec8d9R109) is 5 millisec
- [Infinite loop](https://github.com/takahirom/PreLollipopTransition/pull/18/files#diff-5608e74ca5914ec939cde04c740ec8d9R104) occur if the image saving process hang. I think there should be maximum times to loop, maybe 3 second? not sure.
- [Flag](https://github.com/takahirom/PreLollipopTransition/pull/18/files#diff-ab884a8ff1acc50ce2b9db3d7c0d8bccR45) won't be INIT. (After I send pull request, I felt I don't need INIT state)
- And finally, the flag looks awful. I want to avoid it...but I couldn't find better way to do this.

**Comment by takahirom** [comment]
I agree.
I think using FileLock.
http://developer.android.com/intl/ja/reference/java/nio/channels/FileLock.html
http://inside.ascade.co.jp/node/35

What do you think @shiraji ?

**Comment by shiraji** [comment]
Ah! Good one. I will replace flags with FileLock.

**Comment by shiraji** [comment]
OK, while I designing `FileLock`, I felt something wrong. From what the sample code say, it has to pass `FileOutputStream` to `TransitionAnimation`. Keep opening `OutputStream` often leads memory leak. I really don't want to do that. If closing `FileOutputStream` won't affect `FileLock`, then it would pass `File` object and recreate `FileOutputStream` at `TransitionAnimation`. However, since I have a few (almost none) experience with `FileLock` object, I have no idea how that works.

And then, I changed my mind. Using default multi-thread handling would work well in this case. And it fails circle ci...

**Comment by takahirom** [comment]
> I have a few (almost none) experience with FileLock object, I have no idea how that works.

Sorry me too...
And I think no problem in your way.

New test passed :tada:

**Comment by shiraji** [comment]
yea. should envelope `synchronized`. Fixed the problem.

**Comment by takahirom** [comment]
May I review code?
https://github.com/takahirom/PreLollipopTransition/pull/18/files?w=

**Comment by shiraji** [comment]
Could you wait until I remove [WIP]? (You can review my code anytime but I may change my code/commit until I remove [WIP])

I felt weird for `synchronized` block for `TransitionBundleFactory`. Do you think this code would be better idea? It does not have any blocking code in this class. Nothing to do with `synchronized` block...

``` java
        synchronized (TransitionAnimation.lock) {
            TransitionAnimation.isReady = false;
        }
        new Thread(new Runnable() {
            @Override
            public void run() {
                synchronized (TransitionAnimation.lock) {
                    BufferedOutputStream bos = null;
                    try {
                        if (imageFile.exists()) {
                            imageFile.delete();
                        }
                        imageFile.createNewFile();
                        bos = new BufferedOutputStream(new FileOutputStream(imageFile));
                        bitmap.compress(Bitmap.CompressFormat.JPEG, 50, bos);
                    } catch (FileNotFoundException e) {
                        if (isDebug) {
                            Log.i(TAG, "file not found", e);
                        }
                    } catch (IOException e) {
                        if (isDebug) {
                            Log.i(TAG, "can't create file", e);
                        }
                    } finally {
                        try {
                            bos.close();
                        } catch (Exception e) {
                            if (isDebug) {
                                //IOException, NullPointerException
                                Log.i(TAG, "fail save image", e);
                            }
                        }
                        TransitionAnimation.isReady = true;
                    }
                    TransitionAnimation.bitmapCache = new WeakReference<Bitmap>(bitmap);
                    TransitionAnimation.lock.notify();
                }
            }
        }).start();
```

**Comment by shiraji** [comment]
Forget about my previous comment's code. It fails my local tests. I will remove WIP.

File: pre-lollipop-activity-transition/src/main/java/com/kogitune/activity_transition/core/TransitionAnimation.java [hunk 1]
```diff
@@ -73,11 +76,23 @@ private static void runEnterAnimation(MoveData moveData, TimeInterpolator interp
     private static void setImageToView(View toView, String imageFilePath) {
         Bitmap bitmap;
         if (bitmapCache == null || (bitmap = bitmapCache.get()) == null) {
+            synchronized (lock) {
+                while (isReady) {
```
Comment (takahirom) [inline pre-lollipop-activity-transition/src/main/java/com/kogitune/activity_transition/core/TransitionAnimation.java]: Sorry,  I do not understand..
Why loop while isReady?

I think this will passed.

```
1. saveImage: TransitionAnimation.isReady = false
2. setImageToView synchronized (lock) : not locked, So passed
3. setImageToView while (isReady): isReady is false,So passed
4. setImageToView BitmapFactory.decodeFile: Maybe File not saved yet???
5. saveImage new thread: Image saved
6. saveImage new thread: TransitionAnimation.isReady = true;
7. saveImage new thread: synchronized (TransitionAnimation.lock) 
```

Comment (shiraji) [inline pre-lollipop-activity-transition/src/main/java/com/kogitune/activity_transition/core/TransitionAnimation.java]: This code seems quite hard to read. I will add a commit to explain this code well.

For quick response, it must have `while(!isReady)` inside of synchronize block. I should've put [this line](https://github.com/takahirom/PreLollipopTransition/pull/18/files#diff-ab884a8ff1acc50ce2b9db3d7c0d8bccR43) inside of the thread and enveloped synchronize block. That's why this code seems awkward. This code wants to lock the flag, not its lock object.

**Comment by shiraji** [comment]
Wait...did I sent pull request with `while(isReady)`...this is stupid code. You are absolutely right.

**Comment by takahirom** [comment]
NP :)

**Comment by takahirom** [comment]
mmm...
I suggest using CountDownLatch.
1. saveImage() before start thread: latch = new CowntDownLatch(1);
2. setImageToView() if static field bitmap is null : latch.await(3,TimeUnit.SECONDS)
3. saveImage() new thread: Image saved
4. saveImage() new thread: latch.countDown()
5. setImageToView() if latch.await() is true: BitmapFactory.decodeFile(false means timeout)

I think it is simple.

But, Probably some problem contain this.
- Using static field Object, It sometimes become null.
- I do not think about calling saveImage() method twice.

**Comment by shiraji** [comment]
New idea! 
I haven't have any experience with `CountDownLatch`.  I can learn a lot of new concept. Thanks!

[Sample code](http://www.ibm.com/developerworks/jp/java/library/j-5things5.html#N100B3).

**Comment by shiraji** [comment]
These codes are previous idea. 

I will work on `CountDownLatch` or any other multi-thread handling.

**Comment by takahirom** [comment]
If you think this is the best implementation, you should not overdo it!

**Comment by shiraji** [comment]
I am start thinking this code may be good enough for this function. It satisfies all requirements for this issue. If there is a problem related to this pull request, then we discuss how to overcome the problem. (or you can just fix/refactor it.)

**Comment by takahirom** [comment]
I think mostly ok :)
I will code review.

File: pre-lollipop-activity-transition/src/main/java/com/kogitune/activity_transition/core/TransitionAnimation.java [hunk 2]
```diff
@@ -18,7 +18,10 @@
  */
 public class TransitionAnimation {
     private static final String TAG = "Transition";
+    private static final int MAX_TIME_TO_WAIT = 3000;
+    public static final Object lock = new Object();
     public static WeakReference<Bitmap> bitmapCache;
+    public static boolean isReady = false;
```
Comment (takahirom) [inline pre-lollipop-activity-transition/src/main/java/com/kogitune/activity_transition/core/TransitionAnimation.java]: I think `isReady` variable name is abstract. 
I suggest to use `isImageFileReady`.

File: pre-lollipop-activity-transition/src/main/java/com/kogitune/activity_transition/core/TransitionBundleFactory.java [hunk 3]
```diff
@@ -34,40 +34,49 @@ public static Bundle createTransitionBundle(Context context, View fromView, Bitm
...
-            try {
-                bos.close();
-            } catch (Exception e) {
-                if (isDebug) {
-                    //IOException, NullPointerException
-                    Log.i(TAG, "fail save image", e);
+        new Thread(new Runnable() {
+            @Override
+            public void run() {
+                TransitionAnimation.isReady = false;
```
Comment (takahirom) [inline pre-lollipop-activity-transition/src/main/java/com/kogitune/activity_transition/core/TransitionBundleFactory.java]: mmm.... I think this is flow  below...
Sorry, Probably my understanding is not enough...

| time | 1 | 2 | 3 | 4 | 5 | 6 |
| --- | --- | --- | --- | --- | --- | --- |
| MainThread |  | Synchronized while(!isReady) loop | waiting timed out(3000ms) |  | timeout |  |
| WorkerThread | isReady=false | image writing | isReady = true | can't enter Synchronized block because locked |  | TransitionAnimation.lock.notify(); |

Comment (takahirom) [inline pre-lollipop-activity-transition/src/main/java/com/kogitune/activity_transition/core/TransitionBundleFactory.java]: Sorry, WorkerThread synchronized block is not locked by calling wait().
You are right!

File: pre-lollipop-activity-transition/src/main/java/com/kogitune/activity_transition/core/TransitionAnimation.java (hunk 2 above)
Comment (shiraji) [inline pre-lollipop-activity-transition/src/main/java/com/kogitune/activity_transition/core/TransitionAnimation.java]: Fair enough. I wil fix this.

File: pre-lollipop-activity-transition/src/main/java/com/kogitune/activity_transition/core/TransitionBundleFactory.java (hunk 3 above)
Comment (shiraji) [inline pre-lollipop-activity-transition/src/main/java/com/kogitune/activity_transition/core/TransitionBundleFactory.java]: So, I don't need to do anything, right?

**Comment by shiraji** [comment]
Fixed variable name. If you think this commit is OK, let me know. I will squash commits.

**Comment by shiraji** [comment]
Squashed! (from what @takahirom tweet in Twitter)

**Comment by takahirom** [comment]
Thanks!
I will test manually.

File: pre-lollipop-activity-transition/src/main/java/com/kogitune/activity_transition/core/TransitionBundleFactory.java [hunk 4]
```diff
@@ -34,40 +34,49 @@ public static Bundle createTransitionBundle(Context context, View fromView, Bitm
...
+                        bos.close();
+                    } catch (Exception e) {
+                        if (isDebug) {
+                            //IOException, NullPointerException
+                            Log.i(TAG, "fail save image", e);
+                        }
+                    }
+                    TransitionAnimation.isImageFileReady = true;
+                }
+                TransitionAnimation.bitmapCache = new WeakReference<Bitmap>(bitmap);
```
Comment (takahirom) [inline pre-lollipop-activity-transition/src/main/java/com/kogitune/activity_transition/core/TransitionBundleFactory.java:74]: This static field bitmap cache save timing is maybe too late.
So, now maybe always use file for animation.
Should I this code move to before starting thread?  What do you think @shiraji?

Comment (shiraji) [inline pre-lollipop-activity-transition/src/main/java/com/kogitune/activity_transition/core/TransitionBundleFactory.java:74]: Right. Actually, This is not related to saveImage. We can put outside of saveImage().
